src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Documents/VisualStudioFileChangeTracker.cs

```diff
@@ -72,22 +72,25 @@ public override void StartListening()
         {
             _projectSnapshotManagerDispatcher.AssertDispatcherThread();
 
-            if (_fileChangeUnadviseTask?.IsCompleted == false)
-            {
-                // An unadvise operation is still processing, block the project snapshot manager's thread until it completes.
-                _fileChangeUnadviseTask.Join();
-            }
-
             if (_fileChangeAdviseTask != null)
             {
                 // Already listening
                 return;
             }
 
+            // If an unadvise operation is still processing, we don't start listening until it completes.
+            if (_fileChangeUnadviseTask is not { IsCompleted: false } fileChangeUnadviseTaskToJoin)
+            {
+                fileChangeUnadviseTaskToJoin = null;
+            }
+
             _fileChangeAdviseTask = _joinableTaskContext.Factory.RunAsync(async () =>
             {
                 try
                 {
+                    if (fileChangeUnadviseTaskToJoin is not null)
+                        await fileChangeUnadviseTaskToJoin.JoinAsync().ConfigureAwait(true);
```

Comment:
📝 I'm assuming this is what you want. Behavior should be the same as before except not deadlock.

Reply:
Yup!

@allisonchou do you think we could move the `FileChangeTracker` type off of the dispatcher thread? I don't see it interacting with the `ProjectSnapshotManager` so I thinkkkkkk it'd be ok but please correct me if I'm wrong